internal/sync/lifecycle.go

```diff
@@ -20,24 +20,105 @@
 
 package sync
 
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+// LifecycleState represents `states` that a lifecycle object can be in.
+type LifecycleState int
+
+const (
+	// Idle indicates the Lifecycle hasn't been operated on yet.
+	Idle LifecycleState = iota
+
+	// Starting indicates that the Lifecycle has begun it's "start" command
+	// but hasn't finished yet.
+	Starting
+
+	// Running indicates that the Lifecycle has finished starting and is
+	// available.
+	Running
+
+	// Stopping indicates that the Lifecycle 'stop' method has been called
+	// but hasn't finished yet.
+	Stopping
+
+	// Stopped indicates that the Lifecycle has been stopped.
+	Stopped
+
+	// Errored indicates that the Lifecycle experienced an error and we can't
+	// reasonably determine what state the lifecycle is in.
+	Errored
+)
+
 // LifecycleOnce is a helper for implementing transport.Lifecycles
 // with similar behavior.
 type LifecycleOnce struct {
-	start Once
-	stop  Once
+	lock sync.Mutex
+
+	state     atomic.Int32
+	startOnce sync.Once
+	startErr  error
+	stopOnce  sync.Once
+	stopErr   error
 }
 
 // Start will run the `f` function once and return the error
 func (l *LifecycleOnce) Start(f func() error) error {
-	return l.start.Do(f)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if LifecycleState(l.state.Load()) != Idle {
+		return l.startErr
+	}
+
+	if f == nil {
+		f = func() error { return nil }
+	}
+
+	l.state.Store(int32(Starting))
+	l.startErr = f()
+	if l.startErr == nil {
+		l.state.Store(int32(Running))
+	} else {
+		l.state.Store(int32(Errored))
+	}
```

Comment:
Mostly the reason why state is in an atomic is because though you won't be writing to it at the same time, it is possible to read from it using the `IsRunning` function (or another `State` function we might show)